lib/commands/exam.js

```diff
@@ -254,5 +372,97 @@ module.exports = TestCommand.extend({
+      const browserModuleMapping = this.config.progOptions.browser_module_mapping;
+
+      // Updating browserModuleMapping which maps from browser ids to a list of AMD modules ran per the browser
+      // is needed when running test suite in ci.
+      if (this.config.appMode === 'ci' && browserModuleMapping && !this.moduleQueue) {
+        const browserId = /browser=\s*([0-9]*)/.exec(this.launcher.settings.test_page)[1];
+
+        if (browserId === undefined) {
+          throw new Error('Browser Id can\'t be found.');
+        }
+
+        this.moduleQueue = browserModuleMapping[browserId];
+      } else if (!proto.moduleQueue) {
+        proto.moduleQueue = modules;
+        proto.moduleMap = {};
+      }
+    }
+
+    events['get-next-module'] = function () {
+      const proto = Object.getPrototypeOf(this);
+      const replayExecution = this.config.progOptions.browser_module_mapping;
+      const moduleQueue = this.moduleQueue;
+      const moduleName = moduleQueue.shift();
+
+      if (!moduleName) {
+        this.socket.emit('testem:module-queue-complete');
+      } else {
+        this.socket.emit('testem:next-module-response', moduleName);
+        // keep track of the modules ran per browserId when appMode is ci and replayExecution flag is false
+        if (this.config.appMode === 'ci' && !replayExecution) {
+          const browserId = /browser=\s*([0-9]*)/.exec(this.launcher.settings.test_page)[1];
+
+          if (browserId === undefined) {
+            throw new Error('Browser Id can\'t be found.');
+          }
+
+          proto.moduleMap[browserId] = proto.moduleMap[browserId] || [];
+          proto.moduleMap[browserId].push(moduleName);
+        }
+      }
+
+    }
+
+    if (this.validator.shouldReplayExecution) {
+      return events;
+    }
+
+    // We only need to register `after-tests-complete` event to create json file when the test execution run is not for replay-execution.
+    events['after-tests-complete'] = function () {
+      const proto = Object.getPrototypeOf(this);
+      const browserCount = Object.keys(config.testPage).length;
+      // custom_browser_socket_events is set in config when _loadBalance is in test pages and browser_module_mapping is set when running `replay-execution`.
+      const shouldLoadBalance = !!this.config.progOptions.custom_browser_socket_events && !this.config.progOptions.browser_module_mapping
+
+      proto.finishedBrowsers = proto.finishedBrowsers ? proto.finishedBrowsers + 1 : 1;
+
+      if (proto.finishedBrowsers == browserCount) {
+        // Clean the moduleQueue and finishedBrowser. This guarantees new test run can set the moduleQueue. Otherwise, the server will return no more test to run.
+        // TODO: Investigate not using proto
+        delete proto.moduleQueue;
+        delete proto.finishedBrowser;
+        if (this.config.appMode === 'ci' && shouldLoadBalance) {
+          const fileName = `test-execution-${Date.now()}.json`;
```

Comment:
I realize that this may be unique, but it doesn't help anyone identify an execution easily. Can we instead format this into a more human readable format, and one that will be sortable by time of execution so it's easier to identify a serialized execution that you may want to rerun?